dhall-lsp-server/src/Dhall/LSP/Backend/Completion.hs

```diff
@@ -0,0 +1,186 @@
+completionQueryAt text pos = (completionLeadup, completionPrefix)
+ where
+  off = positionToOffset text pos
+  text' = Text.take off text
+  breakEnd :: (Char -> Bool) -> Text -> (Text, Text)
+  breakEnd p =
+    (\(l,r) -> (Text.reverse l, Text.reverse r)) . Text.break p . Text.reverse
+  (completionPrefix, completionLeadup) =
+    breakEnd (`elem` (" \t\n[(,=+*&|}#?>" :: [Char])) text'
+
+-- | A completion result, optionally annotated with type information.
+data Completion =
+  Completion {
+    completeText :: Text,
+    completeType :: Maybe (Expr Src X) }
+
+-- | Complete file names.
+completeLocalImport :: FilePath -> FilePath -> IO [Completion]
+completeLocalImport relativeTo prefix = do
+  let dir = takeDirectory relativeTo </> takeDirectory prefix
+  exists <- doesDirectoryExist dir
+  if not exists
+  then return []
+  else do
+    files <- listDirectory dir
+    return (map (\file -> Completion (Text.pack file) Nothing) files)
+
+-- | Complete environment variables.
+completeEnvironmentImport :: IO [Completion]
+completeEnvironmentImport = do
+  environment <- getEnvironment
+  let environmentImports = map (Text.pack . fst) environment
+  return $ map (\env -> Completion env Nothing) environmentImports
+
+
+-- | A completion context, consisting of the (approximated) type-checking
+-- context. We need to substitute 'dependent lets' later so we keep their values
+-- around.
+data CompletionContext =
+  CompletionContext {
+    context :: Context (Expr Src X),
+    -- values to be substituted for 'dependent let' behaviour
+    values :: Context (Expr Src X) }
+
+-- | Given a 'binders expression' (with arbitrarily many 'holes') construct the
+-- corresponding completion context.
+buildCompletionContext :: Expr Src X -> CompletionContext
+buildCompletionContext = buildCompletionContext' empty empty
+
+buildCompletionContext' :: Context (Expr Src X) -> Context (Expr Src X)
+  -> Expr Src X -> CompletionContext
+buildCompletionContext' context values (Let (Binding x mA a :| []) e)
+  -- We prefer the actual value over the annotated type in order to get
+  -- 'dependent let' behaviour whenever possible.
+  | Right _A <- typeWithA absurd context a
+  , Right kind <- typeWithA absurd context _A =
+    let _A' = normalize _A
+        kind' = normalize kind
+
+        context' = fmap (shift 1 (V x 0)) $ insert x _A' context
+
+        a' = normalize a
+
+        e' | Const k <- kind', k `elem` [Kind, Sort] =
+             subst (V x 0) a' e  -- 'dependent let'
+           | otherwise = e
+
+        values' | Const k <- kind', k `elem` [Kind, Sort] =
+                  fmap (shift 1 (V x 0)) $ insert x a' values
+                | otherwise = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' e'
+
+  -- fall back to annotated type if body doesn't type check; can't do 'dependent let'
+  | Just _A <- mA
+  , Right _ <- typeWithA absurd context _A =
```

Comment:
I don't think you need this extra type-checking step.  One of the soundness rules of Dhall is that an inferred type like `_A` is also safe to normalize and won't diverge

Reply:
We need this here because the input is not necessarily well-typed. In particular, `_A` may not be a well-formed type.